src/hdr/hdr_decoder.rs

```diff
@@ -0,0 +1,799 @@
+    // There are 8 possible orientations: +Y +X, +X -Y and so on
+    match (c1_tag, c2_tag) {
+        ("-Y", "+X") => {
+            // Common orientation (left-right, top-down)
+            // c1_str is height, c2_str is width
+            let height = try!(c1_str.parse::<u32>().into_image_error(err));
+            let width = try!(c2_str.parse::<u32>().into_image_error(err));
+            Ok((width, height))
+        },
+        _ => {
+            Err(ImageError::FormatError(
+                    format!("Unsupported orientation {} {}", 
+                        limit_string_len(c1_tag, 4), 
+                        limit_string_len(c2_tag, 4))))
+        }
+    } // final expression. Returns value 
+}
+
+trait IntoImageError<T> {
+    fn into_image_error(self, description: &str) -> ImageResult<T>;
+}
+
+impl<T> IntoImageError<T> for ::std::result::Result<T, ::std::num::ParseFloatError> {
+    fn into_image_error(self, description: &str) -> ImageResult<T> {
+        self.map_err(|err| ImageError::FormatError(format!("{} {}", description, err.description())))
+    }
+}
+
+impl<T> IntoImageError<T> for ::std::result::Result<T, ::std::num::ParseIntError> {
+    fn into_image_error(self, description: &str) -> ImageResult<T> {
+        self.map_err(|err| ImageError::FormatError(format!("{} {}", description, err.description())))
+    }
+}
+
+
+// Returns string with no more than len+3 characters
+fn limit_string_len(s: &str, len: usize) -> String {
+    let s_char_len = s.chars().count();
+    if s_char_len > len {
+        s.chars().take(len).chain("...".chars()).collect()
+    } else {
+        s.into()
+    }
+}
+
+// Splits string into (before separator, after separator) tuple
+// or None if separator isn't found 
+fn split_at_first<'a>(s: &'a Cow<'a, str>, separator: &str) -> Option<(&'a str, &'a str)> {
+    match s.find(separator) {
+        None => None,
+        Some(0) => None,
+        Some(p) if p >= s.len()-separator.len() => None,
+        Some(p) => Some((&s[..p], &s[(p+separator.len())..])),
+    } 
+}
+
+#[test]
+fn split_at_first_test() {
+    assert_eq!(split_at_first(&Cow::Owned("".into()), "="), None);
+    assert_eq!(split_at_first(&Cow::Owned("=".into()), "="), None);
+    assert_eq!(split_at_first(&Cow::Owned("= ".into()), "="), None);
+    assert_eq!(split_at_first(&Cow::Owned(" = ".into()), "="), Some((" ", " ")));
+    assert_eq!(split_at_first(&Cow::Owned("EXPOSURE= ".into()), "="), Some(("EXPOSURE"," ")));
+    assert_eq!(split_at_first(&Cow::Owned("EXPOSURE= =".into()), "="), Some(("EXPOSURE"," =")));
+    assert_eq!(split_at_first(&Cow::Owned("EXPOSURE== =".into()), "=="), Some(("EXPOSURE"," =")));
+    assert_eq!(split_at_first(&Cow::Owned("EXPOSURE".into()), ""), None);
+}
+
+// Reads inpult until b"\n" or EOF
+// Returns vector of read bytes NOT including end of line characters
+//   or return None to indicate end of file
+fn read_line_u8<R: Read>(r: &mut R) -> ::std::io::Result<Option<Vec<u8>>> {
```

Comment:
nwin, correct. I can create `BufReader` in `dynimage::load`, then I can pass it to `HDRDecoder<R: BufReader>::new`.

Downside is inconsistency in `dynimage::load` behavior. It's fine to pass `ButReader` into it, while `ImageFormat` is not HDR. 

Well, this can be described in documentation for now. When specialization lands, this special case can be removed.